samcli/local/lambdafn/runtime.py

```diff
@@ -161,28 +242,278 @@ def _get_code_dir(self, code_path):
+        image_builder samcli.local.docker.lambda_image.LambdaImage
+            Instance of the LambdaImage class that can create am image
+        warm_containers bool
+            Determines if the warm containers is enabled or not.
+        """
+        self._containers = {}
+
+        # used to observe any code change in case if warm containers is enabled to support the reloading
+        self._observed_paths = {}
+        self._observed_paths_checksum = {}
+        self._function_paths = {}
+        self._observer = Observer()
+
+        # define matching Event handler, the change event will be fired for any file in all sub directories in the
+        # observed paths. It will also ignore no patterns.
+        self._code_change_handler = PatternMatchingEventHandler(
+            patterns=["*"], ignore_patterns=[], ignore_directories=False
+        )
+        self._code_change_handler.on_modified = self._on_code_change
+        self._lock = threading.Lock()
+        super().__init__(container_manager, image_builder)
+
+    def create(self, function_config, debug_context=None, event=None):
+        """
+        Create a new Container for the passed function, then store it in a dictionary using the function name,
+        so it can be retrieved later and used in the other functions. Make sure to use the debug_context only
+        if the function_config.name equals debug_context.debug-function or the warm_containers option is disabled
+
+        Parameters
+        ----------
+        function_config FunctionConfig
+            Configuration of the function to create a new Container for it.
+        debug_context DebugContext
+            Debugging context for the function (includes port, args, and path)
+        event str
+            input event passed to Lambda function
+
+        Returns
+        -------
+        Container
+            the created container
+        """
+
+        # reuse the cached container if it is created
+        container = self._containers.get(function_config.name, None)
+        if container and container.is_created():
+            LOG.info("Reuse the created warm container for Lambda function '%s'", function_config.name)
+            return container
+
+        # debug_context should be used only if the function name is the one defined
+        # in debug-function option
+        if debug_context and debug_context.debug_function != function_config.name:
+            LOG.debug(
+                "Disable the debugging for Lambda Function %s, as the passed debug function is %s",
+                function_config.name,
+                debug_context.debug_function,
+            )
+            debug_context = None
+
+        container = super().create(function_config, debug_context, None)
+        self._containers[function_config.name] = container
+        self._add_function_to_observer(function_config)
+        return container
+
+    def _on_invoke_done(self, container):
```

Comment:
This function doesn't have any implementation. Is that expected? Or maybe I am miss reading the diff

Reply:
yes, this function is to delete the created container (or to clear any resources) once the invocation is done .. but in warm containers, we can not delete the containers after the invocation is done, we need to wait till the command termination. That is why, I override this function in the WarmContainersRuntime, and make it do nothing.

Reply:
I see now. Thanks for explaining.